wav2vec2 pretraining implemented with speechbrain

RuABraun opened the pull request:
(this is not meant for merge yet but to share the code)

This allows one to pretrain a wav2vec2 model without relying on fairseq or huggingface. It follows the fairseq implementation though there are various differences.

Here is a plot showing WER after finetuning vs different number of pretraining steps and comparing this implementation to fairseq (as well as without and with the quantisation). This is on italian commonvoice 7.0, using validated as the pretraining set, train for finetuning

![pretrain_wer_comparison](https://user-images.githubusercontent.com/12561072/155774998-18b6413b-e700-43ff-857a-4aed224b4b62.png)

I made a wrapper around the whole w2v2 object to make it easy to hold connecting objects like the masking tensor and various projection layers. The feature extractor and encoder are arguments that can be overridden. Using vector quantisation is toggle-able. The implementation follows the most recent fairseq implementation (but not the very recent conformer stuff) so uses normalise_before=True and layer norm is used instead of group norm.

One can use the existing finetuning script inside `/ASR/CTC` to train the pretrained model.

Minor notes and TODOs:
- The queue at my institute favors short jobs so I added a run_opts option "train_time_hours" so I can stop training after x hours (I have another wrapper that then takes care of resubmitting)
- Overloaded fit() for logging
- Overloaded update_average() because original relies on self.step which is reset every epoch plus I don't get the formula why not just use EMA
- Should I actually put the recipe in CommonVoice/self-supervised-learning ? Think some classes could maybe be moved to a better location
- Generally needs more docstrings etc.
- Made a change to make logging output more informative
- I use wandb so there is an option to use that, should change that later to use existing wandb-logger (I saw it but skipped using that for now)

RuABraun commented:
Latest commit is result of merging mine and guille's implementations. Best to ignore commonvoice files as those are not up-to-date and I'd like to focus on getting the librespeech implementation done before we update commonvoice.

TParcollet commented on recipes/LibriSpeech/SSL/hparams/train_wav2vec.yaml:
```diff
@@ -0,0 +1,95 @@
+  embedding_dim: !ref <embedding_dim>
+  final_dim: !ref <final_dim>
+  quantiser: !ref <quantiser>
+  latent_extractor: !ref <extractor>
```
Do we really need to store the extractor and quantiser in the Wav2vec2 class ? I.e can't we compose the forward explicitly in the compute_forward function instead ? If this is not too messy it's usually better.

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+#!/usr/bin/env python3
+"""Recipe for pretraining wav2vec2. See config file for model definition.
+
```
Authors.

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+logger = logging.getLogger(__name__)
+
+
+@sb.utils.checkpoints.register_checkpoint_hooks
```
Why is this needed ?

RuABraun replied:
I was overwriting _save and _recover before, think it can be removed now, will do

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+        padding_mask, mask = padding_mask.to(self.device), mask.to(self.device)
+
+        # normalisation already done in dataloader
+        results = self.modules.model(wavs, wav_lens, padding_mask, mask, is_pretraining=True,
```
See, it might look nice because we only have one line. However, the purpose of compute_forward is to actually expose the forward computation. Again, if this is possible, maybe it would be great to decompose the wav2vec 2 forward here instead.

TParcollet replied:
For instance, we have absolutely no idea what are the outputs because of that.

RuABraun replied:
Yes after thinking about it I completely agree with you! The quantiser is not necessary for inference so it makes more sense to keep it seperate from the wav2vec object I think

gcambara replied:
I'd like to discuss a few things about this.

1. I understand about the advantages of exposing the forward in compute_forward, but at the same time I like the idea that the pretrain forward logic is self-contained within the wav2vec2 class. Someone could simply import our SpeechBrain wav2vec2 module into another framework and easily do the pretrain forward with our "is_pretraining" flag. In any case, I see pros/cons in both approaches, what do you think?

2. Although the quantizer is not needed for inference, I'd suggest keeping it when we release our models, whether it is stored in the wav2vec object or not. I find that it is interesting to enable the community to play around with it, especially for research purposes (I have done explorations on HuggingFace's w2v2 quantizers myself).

TParcollet commented on recipes/LibriSpeech/SSL/hparams/train_wav2vec.yaml:
```diff
@@ -0,0 +1,95 @@
+modules:
+  model: !ref <model>
+
+loss: !new:speechbrain.lobes.models.wav2vec.ContrastiveLoss
```
move the loss to the speechbrain.nnet.losses file ?

RuABraun replied:
done

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+        should_step = self.step % self.grad_accumulation_factor == 0
+        # Managing automatic mixed precision
+        if self.auto_mix_prec:
+            with self.modules.model.no_sync() if not should_step else ExitStack():
```
This I don't understand, what is the matter ?

RuABraun replied:
When doing DDP the gradients are synced on the backward pass, but when doing gradient accumulation every K steps then the syncing is only necessary every K steps and it's wasteful to do it on every backward pass. dan povey pointed it out in this PR https://github.com/speechbrain/speechbrain/pull/1251

the with .. if .. else ExitStack() is syntax so that the `with` statement is only executed when the condition is true

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+                    self.optimizer_step += 1
+
+        if should_step:
+            if isinstance(
```
Why is this needed ?

RuABraun replied:
when wrapping a module with `DistributedDataParallel` one does not have access to the module's attributes anymore, so depending on the whether DDP is used or not the API changes from `normal_module.attribute` to `ddp_module.module.attribute`

It's ugly I agree but don't know of a better approach https://discuss.pytorch.org/t/access-to-attributes-of-model-wrapped-in-ddp/130572

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+                    self.optimizer_step
+                )
+
+            if (
```
Ok I understand the need for logging here. What I propose is to introduce something bigger directly in the core.py of SpeechBrain. I think that logging at every batch totally is relevant for many recipes. What about creating a callback in the official Brain class that is called at the end of fit_batch ? Then you could override it here to manage the logging.

RuABraun replied:
Yes I like that idea (I agree putting logging in fit_batch is awkward), will do

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+                == 0
+            ):
+
+                log_dct = {
```
I guess that this is mandatory for your logging, but it's very hard to follow. Is it mandatory due to wandb?

RuABraun replied:
Do you mean it's unclear because it's not clear why I'm doing this? The idea here is I just log everything (it's not needed because of wandb). Another option is to be more explicit like `logger.info(f'Step: {self.optimizer_step} - loss: {objectives["loss"]} - prob_perplexity: {objectives["prob_perplexity"]} - latent_l2: {objectives["latent_l2"} - lr: {lr} - temp: {temp} ....)`

But that is a lot more verbose and less robust if I want to add or remove something from logging.

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+
+        return objectives["loss"].detach()
+
+    def evaluate_batch(self, batch, stage):
```
why is it redefined ? I don't see a difference with the standard process of Brain.

RuABraun replied:
I think this is something worth discussing, at the moment `evaluate_batch` is assumed to only return one value, but I would actually want to return both the loss and the accuracy (but this breaks because core.py will then try and use the return value to calculate the average loss)

you're right this change at the moment is pointless, when I was overriding `fit()` I had some changes here that I now removed.

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+                    step=self.optimizer_step,
+                )
+
+    def update_average(self, loss, avg_loss):
```
what is this function necessary for ?

RuABraun replied:
The default `update_average` [link](https://github.com/speechbrain/speechbrain/blob/a9b373a9fef3fdfc4e3b4a41478f7b283866186f/speechbrain/core.py#L1239) has two issues for me:

1. it uses `self.step` which is reset every epoch which seems wrong to me (why would you want to have such a reset after every epoch?)
2. it's not clear to me why/how the implementation works

so I used a simple exponential moving average

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+        return avg_loss
+
+
+def sample_negatives(y, num_neg):
```
This is something that could be used by things outside of wav2vec2 and this recipe. What about creating a new file in speechbrain.processing like speechbrain.processing.sequence_management or something like that. I can't think of a better name right now, but this function must be integrated to the SpeechBrain library itself

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+            info.sample_rate, hparams["sample_rate"],
+        )(sig)
+        with torch.no_grad():
+            resampled = F.layer_norm(resampled, resampled.shape)
```
Isn't this a bottleneck cpu-wise ?

RuABraun replied:
will change this is a mistake, only commonvoice needs this

RuABraun replied:
done

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+    train_sampler = DynamicBatchSampler(
+        train_data,
+        hparams["seconds_per_batch"],
+        num_buckets=70,
```
All paremeters -> hparams.yaml :-)

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+        batch_ordering="random",
+    )
+
+    W2VPaddedBatch_partial = functools.partial(W2VPaddedBatch,
```
What is that ?

RuABraun replied:
So I noticed a significant speedup from moving the mask creation into the dataloader. This is what huggingface does and is I suspect the major reason why you find it to be faster than fairseq.

But for the mask creation I need to know some information about the model like the size and number of kernels+strides. So I need to pass that information to a custom collate class (`W2VPaddedBatch`) of the dataloader. `functools.partial` allows me to pass that information to the class without actually initialising it (by setting some specific arguments)

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+    hparams.update(run_opts)
+
+    # Gets hyperparams
+    dct = {
```
What is that ?

RuABraun replied:
for my wandb will remove

TParcollet commented on speechbrain/core.py:
```diff
@@ -266,6 +266,12 @@ def parse_arguments(arg_list=None):
         type=int,
         help="Number of optimizer steps to run. If not passed, all epochs are run.",
     )
+    parser.add_argument(
```
Hum, this must be discussed internally, but I don't see any reason to have this. Stopping with respect to optimisation steps, ok, but not training time.

RuABraun replied:
Will remove this later

TParcollet commented on speechbrain/lobes/augment.py:
```diff
@@ -529,3 +531,60 @@ def _prepare_csv(folder, filelist, csv_file, max_length=None):
         sb.utils.distributed.ddp_barrier()
+
+
+class AddNoiseSimple(torch.nn.Module):
```
More context needed on what this is. Especially since it introduces a new dependency on turtle ?

RuABraun replied:
mistake, was trying some stuff out and forgot to exclude

RuABraun replied:
done

TParcollet commented on speechbrain/lobes/models/transformer/Transformer.py:
```diff
@@ -420,13 +422,16 @@ def __init__(
         )
         self.norm = sb.nnet.normalization.LayerNorm(d_model, eps=1e-6)
+        self.layerdrop_prob = layerdrop_prob
+        self.rng = np.random.default_rng(0)  # for DDP so same layer dropped
```
Replicability here: Should this be set with the seed of the yaml instead ? Always bad to have some constant in the code.

TParcollet commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,610 @@
+from torch.utils.data._utils.collate import default_convert
+from speechbrain.utils.data_utils import mod_default_collate
+from speechbrain.utils.data_utils import recursive_to
+from torch.utils.data._utils.pin_memory import (
```
Why do we need this ?

RuABraun replied:
need for W2VPaddedBatch answered elsewhere

TParcollet commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,610 @@
+        extractor_mode="layer_norm",
+    ):
+        super().__init__()
+        if extractor_mode != "layer_norm":
```
What is the point of having a parameter if we only have one implementation ?

gcambara replied:
Since there are other wav2vec2.0 variants that use Group Norm, we use that argument as a placeholder. But I agree that maybe we could just erase that for simplicity, until we finally implement Group Norm or other variants.

RuABraun replied:
yea this is kind of silly you're right

RuABraun replied:
done

TParcollet commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,610 @@
+
+        latent_l2 = latents.float().pow(2.).mean()
+
+        with torch.cuda.amp.autocast(enabled=False):
```
Why ?

RuABraun replied:
had some issues in the past without, might not be necessary, need to investigate more

TParcollet commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,610 @@
+                    mask_channel_length,
+                )
+            else:
+                with torch.cuda.amp.autocast(enabled=False):
```
Why ?

TParcollet commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,610 @@
+            param.requires_grad = True
+
+    @contextmanager
+    def no_sync(self):
```
?

RuABraun replied:
when wrapped in DDP this method exists. added this method so that when I call it I don't have to do an awkward check for if the module is DDP or not

TParcollet commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,610 @@
+        yield
+
+
+class ContrastiveLoss(nn.Module):
```
Move to speechbrain.nnet.losses

RuABraun replied:
done

TParcollet commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,610 @@
+    return padding_mask
+
+
+def compute_mask(shape, padding_mask, mask_prob, mask_length):
```
Could this also be a more general util function ? I don't think it is specific to w2v2. ?

TParcollet commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,610 @@
+    return mask
+
+
+def apply_mask(
```
The entire masking pipeline is hard to follow, we will need lot of documentation

TParcollet commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,610 @@
+        return hidden_states
+
+
+PaddedData = collections.namedtuple("PaddedData", ["data", "lengths", "padding_mask"
```
?

RuABraun replied:
answered elsewhere

TParcollet commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,610 @@
+    return input_lengths.to(torch.long)
+
+
+class W2VPaddedBatch(PaddedBatch):
```
What is that ?

TParcollet commented on speechbrain/nnet/CNN.py:
```diff
@@ -351,6 +351,7 @@ def __init__(
         bias=True,
         padding_mode="reflect",
         skip_transpose=False,
+        conv_init='',
```
Would, None be better ?

RuABraun replied:
done

TParcollet commented on speechbrain/utils/distributed.py:
```diff
@@ -118,19 +118,19 @@ def ddp_init_group(run_opts):
-                    "Killing process " + str() + "\n"
-                    "Not enough GPUs available!"
-                )
+        # if "local_rank" not in run_opts:
```
?

TParcollet replied:
Still ?

TParcollet commented on speechbrain/utils/logger.py:
```diff
@@ -100,7 +100,7 @@ def setup_logging(
         logging.config.dictConfig(config)
     else:
-        logging.basicConfig(level=default_level)
+        logging.basicConfig(level=default_level, format='%(asctime)s | %(module)s-%(lineno)d | %(levelname)s | %(message)s')
```
Why is this needed ?

RuABraun replied:
By default the output from a message "bla bla" will look like:
```
INFO:root:bla bla
```
With this change a lot more useful information (notice it also prints the module name and line number) is added to the output, now it looks like:
```
2022-04-05 12:14:15,431 | test-4 | INFO | bla bla
```

TParcollet reviewed (changes requested):
Thanks for the huge and awesome work ! Please find my major comments on this review :-)

gcambara commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,398 @@
+#!/usr/bin/env python3
+"""Recipe for pretraining wav2vec2. See config file for model definition.
+
+To run this recipe call python train.py hparams/train_it_wav2vec.yaml --find_unused_parameters --max_grad_norm 0.0
```
```suggestion
To run this recipe call python train.py hparams/train_wav2vec.yaml --find_unused_parameters --max_grad_norm 0.0
```

RuABraun replied:
(going to do these changes locally and then make a big commit)

done.

gcambara commented on recipes/LibriSpeech/librispeech_prepare.py:
```diff
@@ -284,6 +284,9 @@ def create_csv(
+        return
 
     # Preliminary prints
     msg = "Creating csv lists in  %s..." % (csv_file)
```
We should push the symbolic link to "librispeech_prepare.py" under SSL, otherwise this script won't be found by default.

RuABraun replied:
done

gcambara commented on recipes/LibriSpeech/SSL/hparams/train_wav2vec.yaml:
```diff
@@ -0,0 +1,95 @@
+test_csv:
+   - !ref <output_folder>/test-clean.csv
+   - !ref <output_folder>/test-other.csv
+skip_prep: true
```
```suggestion
skip_prep: False
```

Otherwise won't work plug and play.

RuABraun replied:
good catch, done

gcambara commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,610 @@
+        num_layers=8,
+        d_ffn=3072,
+        nhead=8,
+        dropout=0.1,
```
```suggestion
        dropout=0.1,
        activation=nn.GELU,
```
Now the encoder is using ReLU, but the default activation in the wav2vec2.0 paper is GELU.

gcambara commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,610 @@
+            nhead=nhead,
+            d_ffn=d_ffn,
+            d_model=embedding_dim,
+            dropout=dropout,
```
```suggestion
            dropout=dropout,
            activation=nn.GELU,
```

gcambara commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,610 @@
+    def __init__(
+        self,
+        embedding_dim=768,
+        num_layers=8,
```
```suggestion
        num_layers=12,
```
It is the number of layers in the wav2vec2.0 BASE model.

RuABraun commented:
I've refactored a lot, cleaned things up, added more docstrings, think it's worth going through another review.

Some things I still plan to do:
- I think the `W2VLatentEncoder` object is unnecessary, plan to remove and instead just use `TransformerEncoder` directly
- not 100% sure the changes I made to `DynamicBatchSampler` are necessary, need to look into it more
- remove the train_time_hours option I added
- probably should add more info to the README

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,518 @@
+        Boolean mask to mask frames.
+    """
+    wav_lst, latent_length_lst = [], []
+    # If the sample is longer than `max_dur` create several shorter samples from it.
```
Wait, the fact that it will split the data will affect the batch size, correct ? So it bypasses Dynamic Batching, in the sense that the number of batches won't correspond ?

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,518 @@
+    )
+
+
+def _setup_wandb(hparams):
```
Did you try this with multinode training ? If not, we should, because this often break when multinode training is used.

RuABraun replied:
I always do multinode training

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,518 @@
+
+def main():
+    logger.setLevel(logging.INFO)
+    print(sys.argv[1:])
```
remove the print :p

RuABraun replied:
will do

RuABraun replied:
done

TParcollet commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,518 @@
+        },
+    )
+
+    if hparams["use_wandb"]:
```
Can we make wandb compliant with offline mode here ? I mean, 99% of large cluster don't have access to internet on the compute nodes

RuABraun replied:
what do you mean? if this is false then wandb wont be used (so that should be fine for offline mode)

TParcollet commented on speechbrain/core.py:
```diff
@@ -1078,36 +1096,41 @@ def fit(
                             self._save_intra_epoch_ckpt()
                         last_ckpt_time = time.time()
 
-            # Run train "on_stage_end" on all processes
```
What are all these changes on the core.py, I don't understand? This is an absolutely critical place on SB.

TParcollet commented on speechbrain/core.py:
```diff
@@ -1078,36 +1096,41 @@ def fit(
-                        self.on_stage_end,
-                        args=[Stage.VALID, avg_valid_loss, epoch],
-                    )
+            train_time = time.time() - last_ckpt_time
```
This feature shouldn't be part of the PR, please remove :-) We can discuss it elsewhere, but it requires to consult the whole team.

RuABraun replied:
will do

TParcollet commented on speechbrain/core.py:
```diff
@@ -1236,6 +1259,36 @@ def evaluate(
         return avg_test_loss
 
+    @contextmanager
+    def no_sync(self, use=True):
```
Is this affecting the training time? If so, it also looks like something that should be put on an other PR because it affects ALL the recipes and not just this one.

TParcollet commented on speechbrain/dataio/sampler.py:
```diff
@@ -10,6 +10,7 @@
 """
 import torch
```
Ok, could you elaborate on why this is necessary? I remember the issue, but at the same time, we have something that wraps the sampler automatically to prevent this to happen, no ? In which use cases our wrapping isn't sufficient ?

TParcollet commented on speechbrain/lobes/models/convolution.py:
```diff
@@ -133,6 +138,8 @@ def __init__(
                 dilation=dilation,
                 layer_name=f"conv_{i}",
+                bias=conv_bias,
+                padding=padding,
```
I am not sure about the padding. Padding is managed by SpeechBrain, this is why we don't provide it. No ?

RuABraun replied:
This follows fairseq where the input to the feature extractor is not padded. I followed that because I think it makes sense to not bother doing SSL training with representations that are on the "edge".

TParcollet commented on speechbrain/lobes/models/transformer/Transformer.py:
```diff
@@ -420,6 +422,8 @@ def __init__(
         )
         self.norm = sb.nnet.normalization.LayerNorm(d_model, eps=1e-6)
+        self.layerdrop_prob = layerdrop_prob
+        self.rng = np.random.default_rng(0)  # for DDP so same layer(s) are dropped
```
can't we use a given seed instead? Problem is that the same layer will ALWAYS be dropped -> find a way to use self._seed instead ?

RuABraun replied:
Does it really make sense to have a seed variable? It doesn't seem to me like one would ever want to change it (so a constant seems fine to me).

TParcollet commented on speechbrain/lobes/models/wav2vec.py:
```diff
@@ -0,0 +1,315 @@
+import math
```
Function like negative sampling and collate function can go there, if you think that they are only linked to w2v2

RuABraun replied:
okay will move the collate function

TParcollet commented on speechbrain/nnet/losses.py:
```diff
@@ -1128,3 +1128,47 @@ def nll_loss_kd(
     return loss
+
+
+class ContrastiveLoss(nn.Module):
```
Where did you get your inspiration from ? If the code is taken from Fairseq or HF, we must mention it.

RuABraun replied:
will do

RuABraun replied:
done

TParcollet commented on speechbrain/nnet/quantisers.py:
```diff
@@ -0,0 +1,107 @@
+import torch.nn.functional as F
+
+
+class GumbelVectorQuantizer(nn.Module):
```
Where did you get your inspiration from ? If the code is taken from Fairseq or HF, we must mention it.

RuABraun replied:
will do

RuABraun replied:
done

TParcollet commented:
Also please, resolve the conflict so that I can run the tests :p

RuABraun commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,518 @@
+                    kwargs={"data": log_dct, "step": self.optimizer_step,},
+                )
+
+    def evaluate_batch(self, batch, stage):
```
Because accuracy is more meaningful and seems to correspond better to actual performance.
![w2v_loss_acc](https://user-images.githubusercontent.com/12561072/172213436-c58d3390-711c-4ca2-b219-4524dc509a83.png)

In the image above it is the model with the better accuracy (and worse loss) that achieves better WER after finetuning.

RuABraun commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,518 @@
+        return avg_loss
+
+
+def sample_negatives(y, num_neg):
```
will do

RuABraun commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,518 @@
+    return train_loader, valid_loader
+
+
+def w2v_mask_collate_fn(
```
I really think that for this it makes more sense to keep it specific to w2v. We can always move it if another recipe does end up using it.

RuABraun commented on recipes/LibriSpeech/SSL/librispeech_prepare.py:
```diff
@@ -0,0 +1 @@
+../librispeech_prepare.py
```
will do

RuABraun replied:
done

RuABraun commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,518 @@
+            wav_lens.to(self.device),
+            mask.to(self.device),
+        )
+        B = wavs.size(0)
```
will change

RuABraun replied:
done

TParcollet commented:
Hi @RuABraun what are the next steps to follow in your view?

RuABraun commented:
There's the PR #1449 that needs to be merged first. 

I recently figured out one thing that was causing bad performance, but there's still one more issue remaining (get significantly better performance training with 2 GPUs vs with 4). Hope to figure that out in the next ~2 weeks (was busy with interviews in the last ~6 weeks but thankfully that's over now).

Got a few interesting results from trying out different things, will post about them in the slack soon.

RuABraun commented:
waiting on #1518

RuABraun commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,518 @@
+
+        embeddings = results["embeddings"]
+
+        embeddings = embeddings[mask]
```
Can you explain why you think one can't use boolean masks?

RuABraun commented on recipes/LibriSpeech/SSL/train.py:
```diff
@@ -0,0 +1,518 @@
+            and self.optimizer_step % self.hparams.log_interval == 0
+        ):
+
+            log_dct = {
```
done

RuABraun commented:
have updated the PR, before merging I need to attach outputs of a build, think it would be nice if someone else could do the pretraining to confirm it works!

TParcollet commented:
Hi @RuABraun I would remove this logging or put it only on the log.log file (changing the logger). Also, please bring the progress bar back :-) 
<img width="1791" alt="Capture d’écran 2022-08-11 à 17 19 21" src="https://user-images.githubusercontent.com/11910731/184169071-755c230a-52cc-49fc-91b3-e80da518e9f4.png">

RuABraun commented:
@TParcollet  I think the current approach is best for the following reasons.

About the logging:
There are several relevant metrics (loss, accuracy, diversity loss at the minimum) that one wants to track. Is is important for when trying things out to frequently see what these are. Logging these once per epoch is not enough, everybody serious trains on at least thousands of hours, waiting until the end of an epoch to get metrics would take too long.

About the progressbar:
Most people who have the resources to train this will be training on a queue (sungrid engine, slurm etc.). This is the case both in industry and in academia. When you submit jobs on a queue you don't have an interactive commandline interface anymore, instead all the output you would normally see in your terminal will be redirected to a file. Progressbars do not work well with files as output. Additionally, progress-bars cannot display all the different metrics that you care about.

Therefore, I think it is best to have the current behaviour as default.

TParcollet commented:
Hi @RuABraun I hear what you say and partly agree. I think we should keep this intra-step logging, but not in the current form. Can we make it more human-readable i.e. not displaying as a list or dict? Also, note that the progress bar works if you cat the file instead of "vim"-it :p At least, this is what we do on our supercomputer. 

When building a toolkit, it is important to remain consistent, always and forever. All our recipes have progress bars. Now, in this particular case, I agree with you. If we can make the per_step logging more readable, I will be happy with it. Note: we should also bring the epoch_percentage back to the log if the progress bar is removed.

RuABraun commented:
Ah I didn't know about the vim/cat difference, cheers. I like your point about an epoch_percentage!

Not sure how to make it more human-readable though.. I could make it so the decimal numbers are rounded (instead of 0.63242835256...). Not sure if that's enough?

RuABraun commented on speechbrain/lobes/models/transformer/Transformer.py:
```diff
@@ -440,6 +442,8 @@ def __init__(
         )
         self.norm = sb.nnet.normalization.LayerNorm(d_model, eps=1e-6)
+        self.layerdrop_prob = layerdrop_prob
+        self.rng = np.random.default_rng()
```
So we do want this randomness. If the seed were set then the layers dropped will be the same across nodes and that will cause a large drop in performance (notice I do not have a seed in the yaml).

Even if we were to take the rank (node ID) into account while setting the seed, this leads to problems when restarting a build because you are going to be using the same layerdrop and dropout masks as when you started training. Properly setting the seed for DDP training that allows restarting training is not trivial, a hacky solution is possible but I preferred keeping the code clean.

TParcollet replied:
I understand, but a huge point with SB is that everything should be replicable. Can we replace numpy with torch.rand so users don't have to install numpy ?

TParcollet commented:
Rounded numbers and a proper formatted string ? I didn't check but is it also printed in the log.log file ? I guess I could do all this once I am back to work in October:-)

RuABraun commented:
No problem I can do that!

TParcollet commented on recipes/LibriSpeech/ASR/CTC/train_sb_wav2vec.py:
```diff
@@ -0,0 +1,397 @@
+    # NB: This tokenizer corresponds to the one used for the LM!!
+    asr_brain.tokenizer = label_encoder
+
+    pretrained_model_path_extract = (
```
This loading seems a bit unusual compared to what we do. Can't we just only use the checkpointer ? Or PreTrainer ?

TParcollet commented:
Left to be done before merging:
— Finish the training of the BASE on libri and verify perf (add to YAML)
— Upload the model
— Verify the logging of the new refactored version
— Verify pretraining of refactored version
— Verify fine-tuning of refactored version

TParcollet reviewed (approved):
Let's go, only need to test now!

TParcollet commented:
@RuABraun All green! Now, just need to verify that I did not break everything :D

RuABraun commented:
Awesome! :) let's get it finished

RuABraun commented:
Btw I think we could actually just use the scheduler from #1537 as it seems to me it's equivalent.

TParcollet commented:
Well it's a two steps scheduler vs 3 steps. I guess all have reasons to exist ?

RuABraun commented:
Ah nevermind you're right, for some reason I thought it was 3 step.

TParcollet commented:
What an amazing PR! Thank you so much @RuABraun and @gcambara! We are finally done!

RuABraun commented:
@TParcollet Noting two things down here to look at in the future for better performance:

1. No weight decay on biases and layernorm parameters

2. Finetuning on our model just modifies the transformer encoder, the huggingface model finetuning seems to also adjust the convolutional model.
